lib/AST/RequirementMachine/RequirementLowering.cpp

```diff
@@ -455,29 +526,136 @@ void swift::rewriting::realizeInheritedRequirements(
       inferRequirements(inheritedType, loc, moduleForInference, result);
     }
 
-    realizeTypeRequirement(type, inheritedType, loc, result);
+    realizeTypeRequirement(type, inheritedType, loc, result, errors);
   }
 }
 
+bool swift::rewriting::diagnoseRequirementErrors(
+    ASTContext &ctx, SmallVectorImpl<RequirementError> &errors,
+    bool allowConcreteGenericParams) {
+  bool diagnosedError = false;
+  if (ctx.LangOpts.RequirementMachineProtocolSignatures !=
+      RequirementMachineMode::Enabled)
+    return diagnosedError;
+
+  for (auto error : errors) {
+    SourceLoc loc = error.loc;
+    if (!loc.isValid())
+      continue;
+
+    switch (error.kind) {
+    case RequirementError::Kind::InvalidConformance: {
+      Type subjectType = error.requirement.getFirstType();
+      Type constraint = error.requirement.getSecondType();
+
+      // FIXME: The constraint string is printed directly here because
+      // the current default is to not print `any` for existential
+      // types, but this error message is super confusing without `any`
+      // if the user wrote it explicitly.
+      PrintOptions options;
+      options.PrintExplicitAny = true;
+      auto constraintString = constraint.getString(options);
+
+      ctx.Diags.diagnose(loc, diag::requires_conformance_nonprotocol,
+                         subjectType, constraintString);
+      diagnosedError = true;
+
+      auto getNameWithoutSelf = [&](std::string subjectTypeName) {
+        std::string selfSubstring = "Self.";
+
+        if (subjectTypeName.rfind(selfSubstring, 0) == 0) {
+          return subjectTypeName.erase(0, selfSubstring.length());
+        }
+
+        return subjectTypeName;
+      };
+
+      if (allowConcreteGenericParams) {
+        auto subjectTypeName = subjectType.getString();
+        auto subjectTypeNameWithoutSelf = getNameWithoutSelf(subjectTypeName);
+        ctx.Diags.diagnose(loc, diag::requires_conformance_nonprotocol_fixit,
+                           subjectTypeNameWithoutSelf, constraintString)
+             .fixItReplace(loc, " == ");
+      }
+
+      break;
+    }
+
+    case RequirementError::Kind::ConcreteTypeMismatch: {
+      auto type1 = error.requirement.getFirstType();
+      auto type2 = error.requirement.getSecondType();
+
+      if (!type1->hasError() && !type2->hasError()) {
+        ctx.Diags.diagnose(loc, diag::requires_same_concrete_type,
+                           type1, type2);
+        diagnosedError = true;
+      }
+
+      break;
+    }
+
+    case RequirementError::Kind::NonTypeParameter: {
+      ctx.Diags.diagnose(loc, diag::requires_not_suitable_archetype,
```

Comment:
Do we filter out the ErrorType case elsewhere? I imagine it's something like 'DoesNotExist : Int'